src/wms/src/main/java/org/geoserver/wms/map/GetMapKvpRequestReader.java

```diff
@@ -139,78 +175,87 @@ public GetMapKvpRequestReader(WMS wms) {
+        return builder.build();
+    }
+
+    private CacheConfiguration getCacheConfig() {
+        if (null == this.cacheCfg) this.cacheCfg = wms.getRemoteResourcesCacheConfiguration();
+        return this.cacheCfg;
+    }
+
+    EntityResolverProvider getEntityResolverProvider() {
+        if (null == _entityResolverProvider)
+            this._entityResolverProvider = new EntityResolverProvider(wms.getGeoServer());
+        return this._entityResolverProvider;
     }
 
     private int getTimeoutMillis() {
         return wms.getServiceInfo().getRemoteStyleTimeout();
     }
 
-    private synchronized void createHttpClient(
-            WMS wms,
-            HttpClientConnectionManager manager,
-            RequestConfig requestConfig,
-            CacheConfiguration cfg) {
-        this.cacheCfg = cfg;
-        if (cfg != null && cfg.isEnabled()) {
+    private synchronized CloseableHttpClient createHttpClient(
+            RequestConfig requestConfig, CacheConfiguration wmsCacheConfig) {
+        Objects.requireNonNull(requestConfig);
+        Objects.requireNonNull(wmsCacheConfig);
+
+        HttpClientBuilder builder;
+        if (wmsCacheConfig.isEnabled()) {
             CacheConfig cacheConfig =
                     CacheConfig.custom()
-                            .setMaxCacheEntries(cacheCfg.getMaxEntries())
-                            .setMaxObjectSize(cacheCfg.getMaxEntrySize())
+                            .setMaxCacheEntries(wmsCacheConfig.getMaxEntries())
+                            .setMaxObjectSize(wmsCacheConfig.getMaxEntrySize())
                             .build();
-            if (httpClient != null) {
-                try {
-                    httpClient.close();
-                } catch (IOException e) {
-                    if (LOGGER.isLoggable(Level.SEVERE)) {
-                        LOGGER.log(Level.SEVERE, "Error closing HTTPClient", e);
-                    }
-                }
-            }
-            this.httpClient =
+
+            builder =
                     CachingHttpClientBuilder.create()
                             .setCacheConfig(cacheConfig)
-                            .setConnectionManager(manager)
-                            .setDefaultRequestConfig(requestConfig)
-                            .build();
+                            .setConnectionManager(httpClientConnectionManager)
+                            .setDefaultRequestConfig(requestConfig);
         } else {
-            this.httpClient =
-                    HttpClientBuilder.create().setDefaultRequestConfig(requestConfig).build();
+            builder = HttpClientBuilder.create().setDefaultRequestConfig(requestConfig);
+        }
+        return builder.build();
+    }
+
+    private void closeHttpClient() throws IOException {
+        this.cacheCfg = null;
+        @SuppressWarnings("PMD.CloseResource") // closing it right below
+        CloseableHttpClient client = this._httpClient;
```

Comment:
this seems a bit ... convoluted as `this._httpClient;` is defined as a CloseableHttpClient in https://github.com/geoserver/geoserver/pull/5986/files#diff-904a705d3e94b6ec62a0e1d7b2526cb8399ed952e44dcb59be56cd3a879ee360R133

so how about `null` checking and closing that instead of creating a new CloseableHttpClient instance `client` only to call close on that?

Reply:
It's not creating a new CloseableHttpClient instance at all, it's storing the instance variable as a local variable, nullifying the instance variable, and then closing the local variable if it wasn't null to start with. A simple way to not try to close the instance variable twice in a multi-threaded scenario.

Reply:
Though I can see how that can fail too...

Reply:
Switched to good old double checked locking with volatile semantics. please re-review.